### DISCUSSION TITLE
How to build package with multiple libraries

### DISCUSSION THREAD

**Discussion by oraqlle** [vector-of-bool/dds#133] [Q&A] [answered]
When making a package with multiple library roots, DDS currently doesn't generate any build or test any code. The current directory structure looks like so:

```zsh
.
├── CODE_OF_CONDUCT.md
├── CONTRIBUTING.md
├── LICENSE
├── README.md
├── _build
│   ├── _catch-2.10.2
│   │   └── catch2
│   │       ├── catch-main.cpp
│   │       └── catch.hpp
│   ├── _test-driver
│   │   └── v1
│   │       └── catch2
│   │           ├── catch-main.cpp.o
│   │           └── catch-main.cpp.o.d
│   └── compile_commands.json
├── box
│   ├── README.md
│   ├── library.json5
│   └── src
│       ├── box.hpp
│       ├── examples
│       │   └── ...
│       └── tests
│           └── ...
├── concepts
│   ├── README.md
│   ├── library.json5
│   └── src
│       ├── concepts.hpp
│       └── tests
│           ├── arithmatic
│           │   └── ...
│           ├── bitwise
│           │   └── ...
│           └── utility
│               └── ...
├── docs
│   └── ...
├── iterators
│   ├── README.md
│   ├── library.json5
│   └── src
│       ├── adaptors
│       │   └── enumerate.hpp
│       ├── iterators
│       │   ├── column.hpp
│       │   ├── normal.hpp
│       │   ├── reverse.hpp
│       │   ├── row.hpp
│       │   └── two_dim.hpp
│       └── tests
│           ├── adaptors
│           │   └── enumerate
│           │       └── ...
│           └── iterators
│               ├── normal
│               │   └── ...
│               ├── reverse
│               │   └── ...
│               └── two_dim
│                   └── ...
└── package.json5
```
 How do I test and build all the libraries at once. 

Note: These all work, I had them as separate packages before but I'm testing them as a larger package. 

Regards

**Comment by Quincunx271** [comment, marked as answer]
IIRC, box, concepts, and iterators should be under a libs/ directory for
the multi-lib layout.

**Reply by oraqlle** [reply]
Thank you, works like a charm. 

I will also point out that the test cases can't be broken up into separate directories in a multi-lib structure, DDS complains about linking but if you have all your test cases in the bare tests/ directory it works great. 

And, if you are linking against a library in the package to another library in the same package, you can use the `uses: [ ... ]` syntax in `library.json5` of the usee library against the other libraries within the package under the packages namespace. 

Example:

For example, the box library uses both `concepts` and `iterator` This is my current `library.json5` file for box

```json5
{
    name: 'box',
    uses: [
        cortex/iterator,
        cortex/concepts
    ]
}
```

Thought it might be useful for others to point out.
